x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -131,98 +127,95 @@ private void createFollowerIndex(
+        final ActionListener<PutFollowAction.Response> followingListener;
+        if (request.getWaitForCompletion()) {
+            followingListener = listener;
+        } else {
+            followingListener = ActionListener.wrap(() -> {});
+            listener.onResponse(new PutFollowAction.Response(false, false, false));
+        }
 
-                MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
-                IndexMetaData.Builder imdBuilder = IndexMetaData.builder(followIndex);
-
-                // Adding the leader index uuid for each shard as custom metadata:
-                Map<String, String> metadata = new HashMap<>();
-                metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_SHARD_HISTORY_UUIDS, String.join(",", historyUUIDs));
-                metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY, leaderIndexMetaData.getIndexUUID());
-                metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_NAME_KEY, leaderIndexMetaData.getIndex().getName());
-                metadata.put(Ccr.CCR_CUSTOM_METADATA_REMOTE_CLUSTER_NAME_KEY, request.getRemoteCluster());
-                imdBuilder.putCustom(Ccr.CCR_CUSTOM_METADATA_KEY, metadata);
-
-                // Copy all settings, but overwrite a few settings.
-                Settings.Builder settingsBuilder = Settings.builder();
-                settingsBuilder.put(leaderIndexMetaData.getSettings());
-                // Overwriting UUID here, because otherwise we can't follow indices in the same cluster
-                settingsBuilder.put(IndexMetaData.SETTING_INDEX_UUID, UUIDs.randomBase64UUID());
-                settingsBuilder.put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, followIndex);
-                settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
-                settingsBuilder.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
-                imdBuilder.settings(settingsBuilder);
-
-                // Copy mappings from leader IMD to follow IMD
-                for (ObjectObjectCursor<String, MappingMetaData> cursor : leaderIndexMetaData.getMappings()) {
-                    imdBuilder.putMapping(cursor.value);
+        ActionListener<RestoreSnapshotResponse> restoreCompleteHandler = new ActionListener<RestoreSnapshotResponse>() {
+            @Override
+            public void onResponse(RestoreSnapshotResponse restoreSnapshotResponse) {
+                RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+                if (restoreInfo == null || restoreInfo.failedShards() == 0) {
+                    // If restoreInfo is null then it is possible there was a master failure during the
+                    // restore. It is possible the index was restored. initiateFollowing will timeout waiting
+                    // for the shards to be active if the index was not successfully restored.
+                    initiateFollowing(client, request, followingListener);
```

Comment:
this is messy, and we're not really supporting a failover here. What if the current master node dies midway through the restore. Do we resume the follow process on another master? As far as I can see, it's not idempotent (i.e. restore will fail while another restore is already in progress for that index). I would suggest not to handle this case at all for now, and think about it in a follow-up (as well as properly test it).